sky/data/storage.py

```diff
@@ -986,6 +996,66 @@ def sync_all_stores(self):
         for _, store in self.stores.items():
             self._sync_store(store)
 
+    def _maybe_compress_sync_store(self, store: AbstractStore):
+        """Same as sync_store, but compresses before uploading"""
+        # Only supports workdir, self.source should be type str
+        if self.source is None or not isinstance(self.source, str):
+            self._sync_store(store)
+            return
+
+        def num_files(source, excluded_list):
+            count = 0
+            find_cmd = [
+                'find', source, '-type', 'd', '-name', '.git', '-prune', '-o',
+                '-print'
+            ]
+            grep_cmd = ['cat']
+            if len(excluded_list) > 0:
+                grep_cmd = ['grep', '-vE', f'"({"|".join(excluded_list)})"']
+            all_files = subprocess.Popen(find_cmd, stdout=subprocess.PIPE)
+            relevant_files = subprocess.Popen(grep_cmd,
+                                              stdin=all_files.stdout,
+                                              stdout=subprocess.PIPE)
+            all_files.stdout.close()
+            num_files = subprocess.Popen(['wc', '-l'],
+                                         stdin=relevant_files.stdout,
+                                         stdout=subprocess.PIPE)
+            relevant_files.stdout.close()
+            count += int(num_files.communicate()[0])
+            num_files.stdout.close()
+            return count
```

Comment:
As mentioned before, is it possible to use python api instead of subprocesses?

Reply:
I will look into shutil

Reply:
That is a good point! Please benchmark and include some numbers in the discussion ;)

Reply:
100 trials: smaller directory (30 smaller files)
subprocess.run(): 2.9 seconds (modified from Popen, we want the zip process to be blocking)
shutil make_archive: 9.9 seconds

10 trials: large directory (skypilot, 1111 files)
subprocess.run(): 32.4 seconds
shutil make_archive: 46.5 seconds

There is definitely some performance difference, but it's probably overall quite negligible, for the first case that's still only 0.07 seconds difference in a run. The shutil solution is a lot cleaner and easier to understand codewise.

Reply:
Sounds good. Lets keep the subprocess.run()!